llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp

```diff
@@ -320,8 +320,26 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
 
   // abs(X) * abs(X) -> X * X
   Value *X;
-  if (Op0 == Op1 && match(Op0, m_Intrinsic<Intrinsic::abs>(m_Value(X))))
-    return BinaryOperator::CreateMul(X, X);
+  if (Op0 == Op1) {
+    // nabs(X) * nabs(X) -> X * X
+    if (match(Op0, m_Neg(m_Intrinsic<Intrinsic::abs>(m_Value(X))))) {
+      Instruction *NewMul = BinaryOperator::CreateMul(X, X);
+      if (I.hasNoUnsignedWrap()) {
+        // (nabs * nabs) nuw in this case also implies nsw
+        NewMul->setHasNoUnsignedWrap(true);
+        NewMul->setHasNoSignedWrap(true);
+      } else if (I.hasNoSignedWrap())
+        NewMul->setHasNoSignedWrap(true);
+
+      return NewMul;
+    }
```

Comment:
Remove unnecessary nabs handling.

Reply:
nabs actually has a thing that doesn't apply to abs where if it is nuw, the transformation to x * x  is both nuw and nsw.

Reply:
I am aware, but that does not mean that the extra code is useful.

It's okay to improve poison-propagation of an existing transform, but if you want to do more with that, you'll have to provide proof of real-world usefulness.